tensorflow_probability/python/distributions/skew_normal.py

```diff
@@ -0,0 +1,473 @@
+
+  Args:
+    value: Floating-point tensor; the value(s) to be standardized.
+    loc: Floating-point tensor; the location(s) of the distribution(s).
+    scale: Floating-point tensor; the scale(s) of the distribution(s).
+    skewness: Floating-point tensor; the skewness(es) of the distribution(s).
+
+  Returns:
+    A tensor with shape broadcast according to the arguments.
+  """
+  return (value - loc) / tf.math.abs(scale) * tf.math.abs(
+      tf.where(value < loc, x=skewness, y=tf.math.reciprocal(skewness)))
+
+
+def cdf(value, loc, scale, skewness):
+  """Compute cumulative distribution function of Skew Normal distribution.
+
+  Note that scale and skewness can be negative.
+
+  Args:
+    value: Floating-point tensor; where to compute the cdf.
+    loc: Floating-point tensor; the location(s) of the distribution(s).
+    scale: Floating-point tensor; the scale(s) of the distribution(s).
+    skewness: Floating-point tensor; the skewness(es) of the distribution(s).
+
+  Returns:
+    A tensor with shape broadcast according to the arguments.
+  """
+  one = tf.constant(1., dtype=loc.dtype)
+  two = tf.constant(2., dtype=loc.dtype)
+
+  z = standardize(value, loc=loc, scale=scale, skewness=skewness)
+  normal_cdf = tf.cast(special_math.ndtr(z), dtype=loc.dtype)
+
+  squared_skewness = tf.math.square(skewness)
+  return tf.math.reciprocal(one + squared_skewness) * tf.where(
+      z < 0.,
+      x=two * normal_cdf,
+      y=one - squared_skewness + two * squared_skewness * normal_cdf)
+
+
+def quantile(value, loc, scale, skewness):
+  """Compute quantile function (inverse cdf) of Skew Normal distribution.
+
+  Note that scale and skewness can be negative.
+
+  Args:
+    value: Floating-point tensor; where to compute the quantile function.
+    loc: Floating-point tensor; the location(s) of the distribution(s).
+    scale: Floating-point tensor; the scale(s) of the distribution(s).
+    skewness: Floating-point tensor; the skewness(es) of the distribution(s).
+
+  Returns:
+    A tensor with shape broadcast according to the arguments.
+  """
+  half = tf.constant(0.5, dtype=loc.dtype)
+  one = tf.constant(1., dtype=loc.dtype)
+  two = tf.constant(2., dtype=loc.dtype)
+
+  squared_skewness = tf.math.square(skewness)
+  cond = value < tf.math.reciprocal(one + squared_skewness)
+
+  # Here we use the following fact:
+  # X ~ Normal(loc=0, scale=1) => 2 * X**2 ~ Gamma(alpha=0.5, beta=1)
+  probs = (one - value * (one + squared_skewness)) * tf.where(
+      cond, x=one, y=-tf.math.reciprocal(squared_skewness))
+  gamma_quantile = tfp_math.igammainv(half, p=probs)
+
+  abs_skewness = tf.math.abs(skewness)
+  adj_scale = tf.math.abs(scale) * tf.where(
+      cond, x=-tf.math.reciprocal(abs_skewness), y=abs_skewness)
+
+  return loc + adj_scale * tf.cast(
+      tf.math.sqrt(two * gamma_quantile), dtype=loc.dtype)
+
+
+class SkewNormal(distribution.AutoCompositeTensorDistribution):
+  """The Skew Normal distribution.
```

Comment:
I'm a little concerned that we're calling this 'The' skew normal distribution, when there's at least two common variants of this. This one appears to be by the 'type 2' from gamlss, while when people search for skew normal, e.g., on wikipedia they get the 'type 1' (https://en.wikipedia.org/wiki/Skew_normal_distribution). Wikipedia also has https://en.wikipedia.org/wiki/Split_normal_distribution which looks similar to this formulation, but not quite the same. I note you cite `Fernández-Steel distribution` below (I couldn't get a copy of that manuscript, so I don't know the details).

Do you have any insight about what best to call it?

Reply:
Thanks for your feedback!

We can call this “The Two-Piece Normal distribution” [(Robert A. Rigby et al., 2019, Section 18.3.5, p380)][3].

In fact, according to [Rubio and Steel (2020, Section “Who discovered them?”)][4], the two-piece normal distribution “was introduced in a general framework by Fernández and Steel under an alternative parameterisation”. They refer to the formulation proposed by [Fernández and Steel (1998)][2]. So, to be more specific, we can call this distribution “The Two-Piece Normal distribution under the parameterization proposed by [Fernández and Steel (1998)][2]”. And we can name the class `TwoPieceNormal`. What do you think? If you agree, I can make the necessary changes.

We can see the mathematical equivalence between the parameterization proposed by [Fernández and Steel (1998)][2] and the main parameterization discussed in https://en.wikipedia.org/wiki/Split_normal_distribution. Consider that the two-piece normal distribution is built joining at the mode the corresponding halves of two normal distributions with the same mode but different standard deviation. Under the former parameterization, the left (or right) hand side standard deviation is given by `sigma / skewness` (or `sigma * skewness`), where `sigma` and `skewness` are, respectively, the scale and shape parameters of the two-piece normal distribution. Under the latter parameterization, the left (or right) hand side standard deviation is given directly by `sigma_1` (or `sigma_2`). If we use the relationships `sigma_1 = sigma / skewness` and `sigma_2 = sigma * skewness`, then we can go from one parameterization to the other.

The other option consists of maintaining the class name `SkewNormal` but presenting the distribution as "The Fernández-Steel Skew Normal distribution", following [Castillo et al. (2011)][1]. In fact, there is no official version of the skew normal distribution. And the class name `SkewNormal` is better at informing the purpose and main features of this distribution than the class name `TwoPieceNormal` (although the latter is more specific).

[1] Nabor O. Castillo et al. On the Fernández-Steel distribution: Inference and application. _Computational Statistics & Data Analysis_, 55(11), 2951-2961, 2011.

[2] Carmen Fernández and Mark F. J. Steel. On Bayesian modeling of fat tails and skewness. _Journal of the American Statistical Association_, 93(441), 359-371, 1998.

[3] Robert A. Rigby et al. _Distributions for modeling location, scale, and shape: Using GAMLSS in R_. Chapman and Hall/CRC, 2019.

[4] Francisco J. Rubio and Mark F. J. Steel. The family of two‐piece distributions. _Significance_, 17(1), 12-13, 2020.
https://rss.onlinelibrary.wiley.com/doi/full/10.1111/j.1740-9713.2020.01352.x

Reply:
Thanks for looking into it. Let's do this:

1. Rename it to TwoPieceNormal.
2. Add the sentence you proposed (The Two-Piece Normal distribution under the parameterization proposed by [Fernández and Steel (1998)][2])
3. Somewhere in the docstring we should have a section saying that this distribution has also been called the SkewNormal and SplitNormal and cite wikipedia and GAMLSS.

The docstring change should mitigate the somewhat indirect name, so if people seach for skewnormal they might still be able to find this distribution.

Reply:
I made the necessary changes as we agreed.